### PR TITLE
Cooldown support, reporting fixes, + other changes

### PR DESCRIPTION
**Pull request by hauntii**
This PR adds the following to Bandwidth:

- Support for modals and commands to have cooldowns that persist through reboots 
  (Moderator applications have a 30 day cooldown)
- Fix colors for the moderation and piracy warning embeds
- Merge the channel fields on warnings and reports
- Fixes an issue caused by `row` being defined inside an if statement in `messageCreate`, which causes the NLP to crash when force-train is disabled
- Removes `/settings which` in favor of using options instead
- Fix regex bug where SMM level codes are recognized as errors by making sure no letters or numbers are allowed around errors. ( This still allows things like (000-0000) to pass in case someone formats their error that way. )
- Repair user reporting and update the `discord-html-transcript` package
- Allows jumping to offending message from piracy reports
- Moderator applications now notify the user if they have been denied (Accepted users will not be notified)

### CODE REVIEW

**Comment by hauntii** [comment]
@jonbarrow, requesting review as this is getting quite large, and has pretty big fixes.

File: src/context-menus/message/warn-piracy.js [hunk 1]
```diff
@@ -84,7 +79,17 @@ async function warnPiracyHandler(interaction) {
 	});
 	reportEmbed.setTimestamp(Date.now());
 
+	const transcriptButton = new Discord.ButtonBuilder()
+	.setLabel('Jump!')
+	.setStyle(Discord.ButtonStyle.Link)
+	.setEmoji('📨')
+	.setURL(message.url);
```
Comment (jonbarrow) [inline src/context-menus/message/warn-piracy.js:86]: In this codebase we prefer not to chain component functions like this. Instead each `.set` function should be `component.setWhatever()`

File: src/handlers/chat-input-command-handler.js [hunk 2]
```diff
@@ -10,14 +11,25 @@ async function chatInputCommandHandler(interaction) {
...
+	if(!cooldown) 
+	{
+		await command.handler(interaction);
+		if(command.cooldown) { cooldownUtils.beginCooldown(command, memberId) }
+	} else {
+	await interaction.reply({
+		embeds: [cooldown],
+		ephemeral: true
+	});
+	}
```
Comment (jonbarrow) [inline src/handlers/chat-input-command-handler.js:32]: Same formatting comments as before. `if` needs a space, brackets should be on the same line (`if (true) {`), `if` statements are _always_ multi-line, and the interaction reply has bad indentation

File: src/handlers/modal-submit-handler.js [hunk 3]
```diff
@@ -9,14 +11,25 @@ async function modalSubmitHandler(interaction) {
...
+	if(!cooldown) 
+	{
+		await modal.handler(interaction);
+		if(modal.cooldown) { cooldownUtils.beginCooldown(modal, memberId) }
+	} else {
+	await interaction.reply({
+		embeds: [cooldown],
+		ephemeral: true
+	});
+	}
```
Comment (jonbarrow) [inline src/handlers/modal-submit-handler.js:32]: Same formatting comments as before

File: src/modals/mod-application.js [hunk 4]
```diff
@@ -81,9 +81,9 @@ async function modApplicationHandler(interaction) {
 
 	const modApplicationEmbed = new Discord.EmbedBuilder();
 
-	modApplicationEmbed.setColor(0x0096FF);
+	modApplicationEmbed.setColor(0x9D6FF3);
```
Comment (jonbarrow) [inline src/modals/mod-application.js:84]: Why was this color changed?

File: src/modals/report-user.js [hunk 5]
```diff
@@ -58,7 +58,7 @@ async function reportUserHandler(interaction) {
 
 	const reportEmbed = new Discord.EmbedBuilder();
 
-	reportEmbed.setColor(0xC0C0C0);
+	reportEmbed.setColor(0xF36F8A);
```
Comment (jonbarrow) [inline src/modals/report-user.js:61]: Why was this color changed?

File: src/modals/report-user.js [hunk 6]
```diff
@@ -95,26 +90,26 @@ async function reportUserHandler(interaction) {
...
-	const transcriptButton = new Discord.MessageButton();
-	transcriptButton.setEmoji('📜');
-	transcriptButton.setLabel('Download Transcript');
-	transcriptButton.setStyle('LINK');
-	transcriptButton.setURL(message.attachments.first().url);
+	const transcriptButton = new Discord.ButtonBuilder()
+	.setLabel('Download Transcript')
+	.setStyle(Discord.ButtonStyle.Link)
+	.setEmoji('📜')
+	.setURL(message.attachments.first().url);
```
Comment (jonbarrow) [inline src/modals/report-user.js:105]: Same comment as before about not using chaining

File: src/utils/cooldown.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
+const Discord = require('discord.js');
+const database = require('../database');
+
+/**
+ *
+ * @param {Discord.CommandInteraction} command
+ * @param {Number} memberId
+ */
+async function isInteractionOnCooldown(command, memberId) {
+	if(!command.cooldown) {return false}
```
Comment (jonbarrow) [inline src/utils/cooldown.js:10]: Same formatting comments as before

File: src/utils/cooldown.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Initialize our cooldown if not already and grab the cooldown
+	await database.initMemberCooldown(memberId, command.name)
+	const cooldown = await database.getCommandCooldown(memberId, command.name)
+
+    // Check if we are still on cooldown or if there has never been a cooldown for this command yet
+	if(cooldown == 0 || Date.now() > cooldown) {return false}
```
Comment (jonbarrow) [inline src/utils/cooldown.js:17]: Same formatting comments as before

File: src/utils/cooldown.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+async function beginCooldown(command, memberId) {
+	const endTime = (new Date(Date.now() + command.cooldown).getTime());
+	
+	await database.updateCommandCooldown(memberId, command.name, endTime)
+}
+
+function getRelativeTime(timestamp) {
+
+    const msPerMinute = 60 * 1000;
+	const msPerHour = msPerMinute * 60;
```
Comment (jonbarrow) [inline src/utils/cooldown.js:45]: Extra indent here

**Review by jonbarrow — CHANGES_REQUESTED** [review]
Mostly just questions for clarification and formatting changes

File: src/context-menus/message/warn-piracy.js [hunk 10]
```diff
@@ -43,7 +43,7 @@ async function warnPiracyHandler(interaction) {
 
 	const reportEmbed = new Discord.EmbedBuilder();
 
-	reportEmbed.setColor(0xC0C0C0);
+	reportEmbed.setColor(0xF36F8A);
```
[the remark this answers is not in the document]
Comment (hauntii) [reply src/context-menus/message/warn-piracy.js:46]: This color was changed to be more consistent with the actual user-facing piracy message sent.

File: src/modals/mod-application.js (hunk 4 above)
Comment (hauntii) [reply src/modals/mod-application.js:84]: This color was changed to match the accent on the `Pending` banner Bandwidth sends with this message.

File: src/modals/report-user.js (hunk 5 above)
Comment (hauntii) [reply src/modals/report-user.js:61]: This color was changed to be consistent with the other colors changed.

File: src/buttons/mod-application-accept.js [hunk 11]
```diff
@@ -32,7 +32,7 @@ async function modApplicationAcceptHandler(interaction) {
 	const rowOld = message.components[0];
 	const [acceptButtonOld, denyButtonOld] = rowOld.components;
 
-	modApplicationEmbed.setColor(0x2D992D);
+	modApplicationEmbed.setColor(0x6FF38E);
```
[the remark this answers is not in the document]
Comment (hauntii) [reply src/buttons/mod-application-accept.js:35]: This color was changed to match the accent on the `Accepted` banner Bandwidth sends with this message.

**Review by jonbarrow — APPROVED** [review]

**Pull request merged**